controllers/spec/common.go

```diff
@@ -1725,3 +1787,251 @@ func getFilenameOfComponentPackage(componentPackage string) string {
+	return processCommandArgs
+}
+
+func getFunctionRunnerImageDistroless(spec *v1alpha1.FunctionSpec) string {
+	runtime := &spec.Runtime
+	img := spec.Image
+	if img != "" {
+		return img
+	} else if runtime.Java != nil && runtime.Java.Jar != "" {
+		return Configs.RunnerImages.JavaDistroless
+	} else if runtime.Python != nil && runtime.Python.Py != "" {
+		log.Info("Do not support running Python distroless image yet, fall back to use old Python" +
+			" runner image")
+		return Configs.RunnerImages.Python
+	} else if runtime.Golang != nil && runtime.Golang.Go != "" {
+		log.Info("Do not support running Golang distroless image yet, fall back to use old Golang runner image")
+		return Configs.RunnerImages.Go
+	}
+	return DefaultRunnerImage
+}
+
+func getProcessJavaRuntimeArgsDistroless(packageName, logLevel, memory, extraDependenciesDir string) []string {
+	classPath := "/pulsar/instances/java-instance.jar"
+	if extraDependenciesDir != "" {
+		classPath = fmt.Sprintf("%s:%s/*", classPath, extraDependenciesDir)
+	}
+	setLogLevel := ""
+	if logLevel != "" {
+		setLogLevel = strings.Join(
+			[]string{
+				fmt.Sprintf("-Dpulsar.log.level=%s", logLevel),
+				fmt.Sprintf("-Dbk.log.level=%s", logLevel),
+			},
+			" ")
+	}
+	args := []string{
+		"java",
+		"-cp",
+		classPath,
+		fmt.Sprintf("-D%s=%s", FunctionsInstanceClasspath, "/pulsar/lib/*"),
+		fmt.Sprintf("-Dlog4j.configurationFile=%s", DefaultJavaLogConfigPath),
+		"-Dpulsar.function.log.dir=logs/functions",
+		setLogLevel,
+		"-Xmx" + memory,
+		"org.apache.pulsar.functions.instance.JavaInstanceMain",
+		"--jar",
+		packageName,
+		"--config_file",
+		DefaultGenericRunnerConfigFile,
+	}
+
+	return removeEmptyStrings(args)
+}
+
+func makeConfigInitContainerDistroless(configInitCommand string) *corev1.Container {
+
+	volumeMounts := generateConfigInitContainerVolumeMounts()
+	return &corev1.Container{
+		Name:            "config-init-container",
+		Image:           "busybox",
+		Command:         []string{"sh", "-c", configInitCommand},
+		VolumeMounts:    volumeMounts,
+		Env:             generateConfigInitContainerEnv(),
+		ImagePullPolicy: corev1.PullIfNotPresent,
+	}
+}
+
+func getGenerateJavaConfigInitCommandDistroless(clusterName, details, uid string, runtime *v1alpha1.JavaRuntime,
+	authProvided, tlsProvided bool, secretMaps map[string]v1alpha1.SecretRef, state *v1alpha1.Stateful,
+	tlsConfig TLSConfig, authConfig *v1alpha1.AuthConfig, healthCheckInterval int32) string {
+	// this command should do the following things.
+	processCommand := setShardIDEnvironmentVariableCommand() + " && " +
+		generateJavaLogConfigCommandDistroless(runtime) + " " +
+		generateProcessJavaRuntimeConfigFileCommand(
+			clusterName,
+			details,
+			uid,
+			authProvided,
+			tlsProvided,
+			secretMaps,
+			state,
+			tlsConfig,
+			authConfig,
+			healthCheckInterval)
+	return processCommand
+}
+
+func generateConfigInitContainerVolumeMounts() []corev1.VolumeMount {
+	volumeMounts := []corev1.VolumeMount{}
```

Comment:
Generate the volume and volume mounts for the init container. The runner config file will be stored in the volume